Fix bug in SplitCatSimplifier when next_user is an output node

devashishshankar opened the pull request:
Summary:
When simplifying split cat patterns, if next user of a split node was an output node, there was a bug leading to an issue like: P765993221

Basically, the bug was in how args and kwargs of the user were getting replaced, and the code didn't handle nested arg/kwargs.

Using torch.fx.Node functions such as `all_input_nodes` and `replace_input_with` fixes this issue

Differential Revision: D46603618



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @aakhundov @chenyang78

jansel commented on torch/_inductor/fx_passes/pad_mm.py:
```diff
@@ -391,6 +391,8 @@ def pad_bmm(mat1, mat2, m_padded_length, k_padded_length, n_padded_length):
 def _pad_mm_init():
     from .joint_graph import patterns
 
+    counters_ref = counters["inductor"].copy()
```
Perhaps this should go in the lazy_init decorator (and we should use that decorator for this function).

devashishshankar replied:
cc: @eellison

eellison replied:
sgtm

devashishshankar committed:
Fix bug in SplitCatSimplifier when next_user is an output node (#103338)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/103338

When simplifying split cat patterns, if next user of a split node was an output node, there was a bug leading to an issue like: P765993221

Basically, the bug was in how args and kwargs of the user were getting replaced, and the code didn't handle nested arg/kwargs.

Using torch.fx.Node functions such as `all_input_nodes` and `replace_input_with` fixes this issue

Reviewed By: jansel

Differential Revision: D46603618

fbshipit-source-id: 5c6d878dbd2e3361bbba1c14ff900f3d1d93b83b
